src/journal/test/JournalTest.hs

```diff
@@ -78,18 +78,79 @@ appendBSFake bs fj@(FakeJournal bss ix termCount) =
+         else (FakeJournal (Vector.snoc bss bs) ix termCount, Right ())
+    else (fj, Left BackPressure)
+  where
+
+    journalLength :: Int
+    journalLength = sum (Vector.map
+                         (\bs -> align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT) bss)
+    journalLength' :: Int
+    journalLength' = journalLength + align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT
+
+    padding :: ByteString
+    padding = BS.replicate (termLen * termCount - journalLength - hEADER_LENGTH) '0'
+
+    termLen :: Int
+    termLen = oTermBufferLength testOptions
+
+    position = readBytes + unreadBytes
+
+    limit :: Int
+    limit = readBytes + termLen `div` 2
+
+    hardLimit :: Int
+    hardLimit = readBytes + termLen -- TODO: calculate this correctly
+
+    readBytes :: Int
+    readBytes = Vector.sum
+              . Vector.map (\bs -> align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT)
+              . Vector.take (ix - 1)
+              $ bss
+
+    unreadBytes :: Int
+    unreadBytes = Vector.sum
+                . Vector.map (\bs -> align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT)
+                . Vector.drop ix
+                $ bss
+
+appendBSFake' :: ByteString -> FakeJournal -> [(FakeJournal, Either AppendError ())]
+appendBSFake' bs fj@(FakeJournal bss ix termCount) =
+  doTrace
+    (unlines [ "TRACE"
+             , "ix: " ++ show ix
+             , "termCount: " ++ show termCount
+             , "bs: " ++ show (encodeRunLength bs)
+             , "readBytes: " ++ show readBytes
+             , "unreadBytes: " ++  show unreadBytes
+             , "position: " ++ show position
+             , "limit: " ++ show limit
+             , "journalLength': " ++ show journalLength'
+             , "termLen * termCount: " ++ show (termLen * termCount)
+             ]) $
+    if position < limit
+    then noBackPressure
     else if position < hardLimit
-         then [backPressure, noBackPressure]
+         then backPressure : noBackPressure
          else [backPressure]
   where
     backPressure = (fj, Left BackPressure)
     noBackPressure =
       if journalLength' > termLen * termCount
-      then (FakeJournal (if BS.length padding == 0
-                           then bss
-                           else Vector.snoc bss padding) ix (termCount + 1), Left Rotation)
-      else (FakeJournal (Vector.snoc bss bs) ix termCount, Right ())
+      then if journalLength < termLen * termCount
+           then [(FakeJournal (if BS.length padding == 0
+                             then bss
+                             else Vector.snoc bss padding) ix (termCount + 1), Left Rotation)]
+           else [(FakeJournal bss ix (termCount + 1), Left Rotation)]
+      else if journalLength' > termLen * (min 1 (termCount - 1))
+           then [ (FakeJournal (Vector.snoc bss bs) ix termCount, Right ())
+                , (FakeJournal bss ix (termCount + 1), Left Rotation)
```

Comment:
Should it really be a `termCount + 1` here (and same on line 148? I'm note sure I completely understand the different conditions in play here.

Reply:
I think you are right, fixed, thanks!